pennylane/tensorbox/fn.py

```diff
@@ -0,0 +1,416 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Function wrappers for the TensorBox API"""
+# pylint:disable=abstract-class-instantiated
+import warnings
+
+import numpy as np
+
+from .tensorbox import TensorBox
+
+
+def _get_multi_tensorbox(values):
+    """Determines the correct framework to dispatch to given a
+    sequence of tensor like objects.
+
+    Args:
+        values (Sequence[tensor_like]): a sequence of tensor like objects
+
+    Returns:
+        .TensorBox: A TensorBox that will dispatch to the correct framework
+        given the rules of precedence. This TensorBox will contain the *first*
+        tensor like object in ``values`` that corresponds to the highest-priority
+        framework.
+
+    To determine the framework to dispatch to, the following rules
+    are followed:
+
+    * Tensors that are incompatible (such as Torch and TensorFlow tensors)
+      cannot both be present.
+
+    * Autograd tensors *may* be present alongside Torch and TensorFlow tensors,
+      but Torch and TensorFlow take precendence; the autograd arrays will
+      be treated as non-differentiable NumPy arrays. A warning will be raised
+      suggesting that vanilla NumPy be used instead.
+
+    * Vanilla NumPy arrays can be used alongside other tensor objects; they will
+      always be treated as non-differentiable constants.
+    """
+    interfaces = [get_interface(v) for v in values]
+
+    if "tf" in interfaces and "torch" in interfaces:
+        raise ValueError("Tensors contain mixed types; cannot determine dispatch library")
+
+    non_numpy_interfaces = set(interfaces) - {"numpy"}
+
+    if len(non_numpy_interfaces) == 2:
+        warnings.warn(
+            f"Contains tensors of types {non_numpy_interfaces}; dispatch will prioritize "
+            "TensorFlow and PyTorch over autograd. Consider replacing Autograd with vanilla NumPy.",
+            UserWarning,
+        )
+
+    if "tf" in interfaces:
+        return TensorBox(values[interfaces.index("tf")])
+
+    if "torch" in interfaces:
+        return TensorBox(values[interfaces.index("torch")])
+
+    if "autograd" in interfaces:
+        return TensorBox(values[interfaces.index("autograd")])

```

Comment:
so this function takes a seq of objects, and joins only the objects of the priority interface to a new tensorbox? the other ones get dropped? isn't that unexpected?

Reply:
Good question.

The idea here was not so much to return a wrapped tensorbox for every value in the sequence; we just need a _single_ tensorbox so that we know what interface to use. Once we have the right tensorbox, it's just a matter of static method chaining.

So the value of the returned tensorbox isn't really important; it could even be empty. Perhaps worth having this function return an 'empty' tensorbox, as long as it dispatches to the correct framework?